File: ts/src/suggestion_engine/rules/cross_cutting.ts
import { SEVERITY_ERROR, SEVERITY_WARNING } from "../../reviewer/types.ts";
import type { SuggestionRule } from "../protocol.ts";
import type { Suggestion } from "../types.ts";
import type { EvalResult } from "../../types.ts";

export class RegressionRule implements SuggestionRule {
  private readonly threshold: number;

  constructor(regressionThreshold = 0.05) {
    this.threshold = regressionThreshold;
  }

  evaluate(
    result: EvalResult,
    _verdicts?: unknown,
    context?: Record<string, unknown>,
  ): Suggestion[] {
    if (!context || !context["baseline"]) return [];

    const baseline = context["baseline"] as EvalResult[];
    const baselineForCategory = baseline.find(
      (b) => b.category === result.category,
    );
    if (!baselineForCategory) return [];

    const suggestions: Suggestion[] = [];
    const metrics: Array<{ key: keyof EvalResult["metrics"]; name: string }> = [
      { key: "pAt5", name: "p_at_5" },
      { key: "rAt10", name: "r_at_10" },
    ];

    for (const { key, name } of metrics) {
      const current = result.metrics[key] as number;
      const base = baselineForCategory.metrics[key] as number;
      if (typeof current !== "number" || typeof base !== "number") continue;
      const delta = current - base;
      if (delta < -this.threshold) {
        suggestions.push({
          severity: SEVERITY_ERROR,
          message: `${name} regressed: ${base.toFixed(3)} -> ${current.toFixed(3)} (delta: ${delta >= 0 ? "+" : ""}${delta.toFixed(3)})`,
          ruleName: "regression",
          metric: name,
          category: result.category,
          fixHint:
            "Compare recent changes against the baseline run to identify the cause.",
        });
      }
    }
    return suggestions;
  }
}

export class CategoryImbalanceRule implements SuggestionRule {
  private readonly ratio: number;

  constructor(imbalanceRatio = 3.0) {
    this.ratio = imbalanceRatio;
  }

  evaluate(
    _result: EvalResult,
    _verdicts?: unknown,
    context?: Record<string, unknown>,
  ): Suggestion[] {
    if (!context || !context["all_results"]) return [];

    const allResults = context["all_results"] as EvalResult[];
    const counts = allResults
      .filter((r) => r.metrics.n > 0)
      .map((r) => r.metrics.n);

    if (counts.length < 2) return [];

    const maxN = Math.max(...counts);
    const minN = Math.min(...counts);

    if (minN === 0 || maxN / minN <= this.ratio) return [];

    return [
      {
        severity: SEVERITY_WARNING,
        message: `Category fixture counts are imbalanced: smallest=${minN}, largest=${maxN} (ratio ${(maxN / minN).toFixed(1)}x)`,
        ruleName: "category_imbalance",
        metric: "fixture_count",
        fixHint:
          "Add more fixtures to underrepresented categories for balanced evaluation.",
      },
    ];
  }
}

File: src/synapt_eval/suggestion_engine/rules/cross_cutting.py
"""Cross-cutting suggestion rules."""

from __future__ import annotations

from typing import Any

from synapt_eval.reviewer.types import SEVERITY_ERROR, SEVERITY_WARNING, Verdict
from synapt_eval.runner.orchestration import compute_deltas
from synapt_eval.suggestion_engine.types import Suggestion
from synapt_eval.types import EvalResult


class RegressionRule:
    """Flags metric regressions compared to a baseline.

    Consumes baseline results from context["baseline"]. If no
    baseline is provided, the rule produces no suggestions.
    """

    def __init__(self, regression_threshold: float = 0.05) -> None:
        self._threshold = regression_threshold

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not context or "baseline" not in context:
            return []

        baseline: list[EvalResult] = context["baseline"]
        deltas = compute_deltas([result], baseline, self._threshold)
        regressions = [d for d in deltas if d.regression]

        if not regressions:
            return []

        suggestions: list[Suggestion] = []
        for d in regressions:
            suggestions.append(
                Suggestion(
                    severity=SEVERITY_ERROR,
                    message=(
                        f"{d.metric} regressed: {d.baseline:.3f} -> {d.current:.3f} "
                        f"(delta: {d.delta:+.3f})"
                    ),
                    rule_name="regression",
                    metric=d.metric,
                    category=d.category,
                    fix_hint=(
                        "Compare recent changes against the baseline run to identify the cause."
                    ),
                )
            )
        return suggestions


class CategoryImbalanceRule:
    """Flags when fixture coverage is uneven across categories.

    Operates across multiple results via context["all_results"].
    If the ratio between smallest and largest category exceeds the
    imbalance threshold, a suggestion is generated.
    """

    def __init__(self, imbalance_ratio: float = 3.0) -> None:
        self._ratio = imbalance_ratio

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not context or "all_results" not in context:
            return []

        all_results: list[EvalResult] = context["all_results"]
        counts = [r.metrics.n for r in all_results if r.metrics.n > 0]

        if len(counts) < 2:
            return []

        max_n = max(counts)
        min_n = min(counts)

        if min_n == 0 or max_n / min_n <= self._ratio:
            return []

        return [
            Suggestion(
                severity=SEVERITY_WARNING,
                message=(
                    f"Category fixture counts are imbalanced: "
                    f"smallest={min_n}, largest={max_n} (ratio {max_n / min_n:.1f}x)"
                ),
                rule_name="category_imbalance",
                metric="fixture_count",
                fix_hint=(
                    "Add more fixtures to underrepresented categories for balanced evaluation."
                ),
            )
        ]
